library/pem.c

```diff
@@ -158,99 +158,63 @@ static int pem_pbkdf1(unsigned char *key, size_t keylen,
+    size_t output_len;
+    int ret;
+
+    cipher_info = mbedtls_cipher_info_from_type(enc_alg);
+    if (cipher_info == NULL) {
+        return MBEDTLS_ERR_PEM_FEATURE_UNAVAILABLE;
     }
-    ret = mbedtls_des_crypt_cbc(&des_ctx, MBEDTLS_DES_DECRYPT, buflen,
-                                des_iv, buf, buf);
-
-exit:
-    mbedtls_des_free(&des_ctx);
-    mbedtls_platform_zeroize(des_key, 8);
-
-    return ret;
-}
 
-/*
- * Decrypt with 3DES-CBC, using PBKDF1 for key derivation
- */
-static int pem_des3_decrypt(unsigned char des3_iv[8],
-                            unsigned char *buf, size_t buflen,
-                            const unsigned char *pwd, size_t pwdlen)
-{
-    mbedtls_des3_context des3_ctx;
-    unsigned char des3_key[24];
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    key_bitlen = mbedtls_cipher_info_get_key_bitlen(cipher_info);
+    block_size = mbedtls_cipher_info_get_block_size(cipher_info);
 
-    mbedtls_des3_init(&des3_ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        return ret;
+    }
 
-    if ((ret = pem_pbkdf1(des3_key, 24, des3_iv, pwd, pwdlen)) != 0) {
+    ret = pem_pbkdf1(key, key_bitlen/8, iv, pwd, pwdlen);
+    if (ret != 0) {
         goto exit;
     }
 
-    if ((ret = mbedtls_des3_set3key_dec(&des3_ctx, des3_key)) != 0) {
+    ret = mbedtls_cipher_setkey(&ctx, key, (int) key_bitlen, MBEDTLS_DECRYPT);
+    if (ret != 0) {
         goto exit;
     }
-    ret = mbedtls_des3_crypt_cbc(&des3_ctx, MBEDTLS_DES_DECRYPT, buflen,
-                                 des3_iv, buf, buf);
 
-exit:
-    mbedtls_des3_free(&des3_ctx);
-    mbedtls_platform_zeroize(des3_key, 24);
+    output_len = input_len + block_size;
+    output = mbedtls_calloc(1, output_len);
 
-    return ret;
-}
-#endif /* MBEDTLS_DES_C */
-
-#if defined(MBEDTLS_AES_C)
-/*
- * Decrypt with AES-XXX-CBC, using PBKDF1 for key derivation
- */
-static int pem_aes_decrypt(unsigned char aes_iv[16], unsigned int keylen,
-                           unsigned char *buf, size_t buflen,
-                           const unsigned char *pwd, size_t pwdlen)
-{
-    mbedtls_aes_context aes_ctx;
-    unsigned char aes_key[32];
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-
-    mbedtls_aes_init(&aes_ctx);
-
-    if ((ret = pem_pbkdf1(aes_key, keylen, aes_iv, pwd, pwdlen)) != 0) {
+    ret = mbedtls_cipher_crypt(&ctx, iv, iv_len, input, input_len,
+                               output, &output_len);
+    if (ret != 0) {
         goto exit;
     }
 
-    if ((ret = mbedtls_aes_setkey_dec(&aes_ctx, aes_key, keylen * 8)) != 0) {
-        goto exit;
-    }
-    ret = mbedtls_aes_crypt_cbc(&aes_ctx, MBEDTLS_AES_DECRYPT, buflen,
-                                aes_iv, buf, buf);
+    memcpy(input, output, input_len);
 
 exit:
-    mbedtls_aes_free(&aes_ctx);
-    mbedtls_platform_zeroize(aes_key, keylen);
-
+    mbedtls_cipher_free(&ctx);
+    if (output != NULL) {
+        mbedtls_free(output);
+    }
```

Comment:
The content of the `key` variable must be erased from the memory with `mbedtls_platform_zeroize()`.